File: practice.hs
{-# LANGUAGE NoImplicitPrelude #-}
-- the language pragma above allows us to redefine functions
-- that are already defined in prelude

import Prelude (
  Ord, Show, Bool, Int, String, show,
  (<), (>=), (+), (-), (*), div, (++))
import Data.List (sort)

foldr :: (a -> b -> b) -> b -> [a] -> b
foldr f acc [] = acc
foldr f acc (x:xs) = f x (foldr f acc xs)

foldl :: (b -> a -> b) -> b -> [a] -> b
foldl f acc [] = acc
foldl f acc (x:xs) = foldl f (f acc x) xs

map :: (a -> b) -> [a] -> [b]
map f xs = foldr (\x acc -> (f x):acc) [] xs

filter :: (a -> Bool) -> [a] -> [a]
filter f xs = foldr (\x acc -> if f x then x:acc else acc) [] xs

length :: [a] -> Int
length xs = foldr (\x acc -> acc + 1) 0 xs

reverse' :: [a] -> [a]
reverse' xs = foldl (\acc x -> x:acc) [] xs

data List a = Cons a (List a) | Nil

sum :: List Int -> Int
sum Nil = 0
sum (Cons n tl) = n + sum tl

data BTree a =
    Node a (BTree a) (BTree a)
  | Leaf a
  | NilLeaf

preOrder :: BTree a -> [a]
preOrder (Leaf x)  = [x]
preOrder (Node x left right) =
  [x] ++ (preOrder left) ++ (preOrder right)

data Arith =
    Val Int
  | Add Arith Arith
  | Sub Arith Arith
  | Mul Arith Arith

serializeArith :: Arith -> String
serializeArith (Val n) = "int\n" ++ show n
serializeArith (Add x y) = "add\n" ++ serializeArith x ++ serializeArith y
serializeArith (Sub x y) = "sub\n" ++ serializeArith x ++ serializeArith y
serializeArith (Mul x y) = "mul\n" ++ serializeArith x ++ serializeArith y

eval :: Arith -> Int
eval (Val n)   = n
eval (Add x y) = (eval x) + (eval y)
eval (Sub x y) = (eval x) - (eval y)
eval (Mul x y) = (eval x) * (eval y)

data Maybe a = Just a | Nothing

head :: [a] -> Maybe a
head []     = Nothing
head (x:xs) = Just x

insertSort :: Ord a => [a] -> [a]
-- add elements one by one to acc, which is kept sorted
insertSort xs = foldr insert [] xs
        -- insert at the end of sorted sublist
	where insert x [] = [x]
        -- insert at current position or recurse to rest of list
	      insert x (y:ys) = if x >= y then y:(insert x ys) else x:y:ys

mergeSort :: Ord a => [a] -> [a]
mergeSort [] = []
mergeSort [x] = [x]
mergeSort l =
  -- merge sorted sublists
  merge (mergeSort left) (mergeSort right)
        -- split into two sublists
  where (left, right) = splitAt (length l `div` 2) l
        merge [] [] = []
        -- right list is empty; rest of left is end of sorted list
        merge xs [] = xs
        -- left list is empty; rest of right is end of sorted list
        merge [] ys = ys
        -- pick lower head and recurse on the rest
        merge (x:xs) (y:ys) =
          if x < y then x:(merge xs (y:ys))
                   else y:(merge (x:xs) ys)

quickSort :: Ord a => [a] -> [a]
quickSort [] = []
quickSort (x:xs) =
  -- append into complete sorted list
  (quickSort left) ++ [x] ++ (quickSort right)
  where left = filter (x >=) xs -- get lower/eq sublist
        right = filter (x <) xs -- get higher sublist

quickSort2 :: Ord a => [a] -> [a]
quickSort2 [] = []
quickSort2 (x:xs) =
  -- append into complete sorted list
  (quickSort2 left) ++ [x] ++ (quickSort2 right)
  where (left, right) = partition xs -- create partition
        -- repeated put list elems into right sublist
        partition = foldr part ([], [])
        -- put element in the right sublist
        part y (low, hi) =
          if x >= y then (y:low, hi) else (low, y:hi)

append :: [a] -> [a] -> [a]
append [] ys = ys
append (x:xs) ys = x:(append xs ys)

reverse :: [a] -> [a]
reverse [] = []
reverse (x:xs) = insertBack x (reverse xs)
  where insertBack x [] = [x]
        insertBack x (y:ys) = y:(insertBack x ys)

treeSum :: BTree Int -> Int
treeSum NilLeaf = 0
treeSum (Leaf x) = x
treeSum (Node x left right) =
  x + (treeSum left) + (treeSum right)

unzip :: [(a,b)] -> ([a], [b])
unzip tups =
  foldr (\(x,y) (xs,ys) -> (x:xs, y:ys)) ([], []) tups

splitAt :: Int -> [a] -> ([a], [a])
splitAt n xs = splitAt_ n xs []
  where splitAt_ 0 ys acc     = (acc, ys)
        splitAt_ n [] acc     = (acc, [])
        splitAt_ n (y:ys) acc =
          splitAt_ (n-1) ys (acc ++ [y])

File: practice.ml
(*
 * practice.ml
 * code from functional programming practice session
 *)

let rec fold_right (f:'a -> 'b -> 'b) (l:'a list) (acc:'b) : 'b =
  match l with
    [] -> acc
  | x::xs -> f x (fold_right f xs acc)
;;

let rec fold_left (f:'b -> 'a -> 'b) (acc:'b) (l:'a list) : 'b =
  match l with
    [] -> acc
  | x::xs -> fold_left f (f acc x) xs
;;

let map (f:'a -> 'b) (xs:'a list) : 'b list =
  List.fold_right (fun x acc -> (f x)::acc) xs [] ;;

let filter (f:'a -> bool) (xs:'a list) : 'a list =
  List.fold_right (fun x acc -> if f x then x::acc else acc) xs [] ;;

let length (xs:'a list) : int =
  List.fold_right (fun x acc -> acc + 1) xs 0 ;;

let reverse (xs:'a list) : 'a list =
  List.fold_left (fun acc x -> x::acc) [] xs ;;

type 'a list2 =
  | Cons of 'a * ('a list2)
  | Nil
;;

let rec sum (l:int list2) : int =
  match l with
    Nil -> 0
  | Cons (x, tl) -> x + sum tl
;;

type 'a btree =
    Node of 'a * ('a btree) * ('a btree)
  | Leaf of 'a
  | NilLeaf
;;

let rec preOrder (tree:'a btree) : 'a list = 
  match tree with
    NilLeaf -> []
  | Leaf x -> [x]
  | Node (x, left, right) ->
      [x] @ (preOrder left) @ (preOrder right)
;;

type arith =
  | Val of int
  | Add of arith * arith
  | Sub of arith * arith
  | Mul of arith * arith
;;

let rec serializeArith (ast:arith) : string =
  match ast with
    Val n -> "int\n" ^ (string_of_int n) ^ "\n"
  | Add (x,y) -> "add\n" ^ serializeArith x ^ serializeArith y
  | Sub (x,y) -> "sub\n" ^ serializeArith x ^ serializeArith y
  | Mul (x,y) -> "mul\n" ^ serializeArith x ^ serializeArith y
;;

let rec eval (a:arith) : int =
  match a with
    Val n -> n
  | Add (x,y) -> (eval x) + (eval y)
  | Sub (x,y) -> (eval x) - (eval y)
  | Mul (x,y) -> (eval x) * (eval y)
;;

let insertSort (l:int list) : int list =
  let rec insert x il = begin
    match il with
      [] -> [x]
    | y::ys -> if x >= y then y::(insert x ys) else x::y::ys
  end in
  List.fold_right insert l []
;;

let splitAt (n:int) (l:'a list) : 'a list * 'a list =
  let rec splitAt_ n2 l2 acc = begin
    match (n2,l2) with
      (0, ys) -> (acc, ys)
    | (n, []) -> (acc, [])
    | (n, y::ys) -> splitAt_ (n-1) ys (acc @ [y])
  end in
  splitAt_ n l []
;;

type 'a option =
  | Some of 'a
  | None
;;

let head (l:'a list) : 'a option =
  match l with
    [] -> None
  | x::xs -> Some x
;;

let insertSort (l:int list) : int list =
  let rec insert x il = begin
    match il with
    (* insert at the end of sorted sublist *)
      [] -> [x]
    (* insert at current position or recurse to rest of list *)
    | y::ys -> if x >= y then y::(insert x ys) else x::y::ys
  end in
  (* add elements one by one to acc, which is kept sorted *)
  List.fold_right insert l []
;;

let rec mergeSort (l:int list) : int list =
  let rec merge xxs yys = begin
    match (xxs,yys) with
      ([], []) -> []
    (* right list is empty; rest of left is end of sorted list *)
    | (xs, []) -> xs
    (* left list is empty; rest of right is end of sorted list *)
    | ([], ys) -> ys
    (* pick lower head and recurse on the rest *)
    | (x::xs, y::ys) ->
        if x < y then x::(merge xs (y::ys))
                 else y::(merge (x::xs) ys)
  end in
  match l with
    [] -> []
  | [x] -> [x]
  | _ ->
    (* split into two sublists *)
    let (left, right) = splitAt (List.length l / 2) l in
    (* merge sorted sublists *)
    merge (mergeSort left) (mergeSort right)
;;

let rec quickSort (l:int list) : int list =
  match l with
    [] -> []
  | x::xs ->
    (* get lower/eq sublist *)
    let left = List.filter (fun y -> x >= y) xs in
    (* get higher sublist *)
    let right = List.filter (fun y -> x < y) xs in
    (* append into complete sorted list *)
    (quickSort left) @ [x] @ (quickSort right)
;;

let rec quickSort2 (l:int list) : int list =
  match l with
    [] -> []
  | x::xs ->
    (* put element in the right sublist *)
    let part y (lo, hi) = if x >= y then (y::lo, hi) else (lo, y::hi) in
    (* repeated put list elems into right sublist *)
    let partition lp = List.fold_right part lp ([],[]) in
    (* create partition *)
    let (left, right) = partition xs in
    (* append into complete sorted list *)
    (quickSort2 left) @ [x] @ (quickSort2 right)
;;

let rec append (xxs:'a list) (yys:'a list) : 'a list =
  match xxs with
    [] -> yys
  | x::xs -> x::(append xs yys)
;;

let rec reverse (l:'a list) : 'a list =
  let rec insertBack y xxs = begin
    match xxs with
      [] -> [y]
    | x::xs -> x::(insertBack y xs)
  end in
  match l with
    [] -> []
  | x::xs -> insertBack x (reverse xs)
;;

let rec treeSum (tree:int btree) : int =
  match tree with
    NilLeaf -> 0
  | Leaf n -> n
  | Node (n,left,right) -> n + (treeSum left) + (treeSum right)
;;

let unzip (l:('a * 'b) list) : ('a list) * ('b list) =
  List.fold_right (fun (x,y) (xs,ys) -> (x::xs, y::ys)) l ([], []) ;;

let splitAt (n:int) (l:'a list) : ('a list) * ('a list) =
  let rec splitAt_ n2 l2 acc = begin
    match (n2,l2) with
      (0, ys) -> (acc, ys)
    | (n, []) -> (acc, [])
    | (n, y::ys) -> splitAt_ (n-1) ys (acc @ [y])
  end in
  splitAt_ n l []
;;
